internal/conf/jwk.go

```diff
@@ -0,0 +1,192 @@
+		if err != nil {
+			return err
+		}
+		pubJwk, err := jwk.PublicKeyOf(privJwk)
+		if err != nil {
+			return err
+		}
+
+		// all public keys will be used for signature verification
+		if pubJwk.KeyUsage() == "enc" {
+			if err := pubJwk.Set(jwk.KeyUsageKey, "sig"); err != nil {
+				return err
+			}
+		}
+		config[pubJwk.KeyID()] = JwkInfo{
+			PublicKey:  pubJwk,
+			PrivateKey: privJwk,
+		}
+	}
+	*j = config
+	return nil
+}
+
+func (j *JwtKeysDecoder) Validate() error {
+	// Validate performs _minimal_ checks if the data stored in the key are valid.
+	// By minimal, we mean that it does not check if the key is valid for use in
+	// cryptographic operations. For example, it does not check if an RSA key's
+	// `e` field is a valid exponent, or if the `n` field is a valid modulus.
+	// Instead, it checks for things such as the _presence_ of some required fields,
+	// or if certain keys' values are of particular length.
+	//
+	// Note that depending on th underlying key type, use of this method requires
+	// that multiple fields in the key are properly populated. For example, an EC
+	// key's "x", "y" fields cannot be validated unless the "crv" field is populated first.
+	for _, key := range *j {
+		if err := key.PrivateKey.Validate(); err != nil {
+			return err
+		}
+		// symmetric keys don't have public keys
+		if key.PublicKey != nil {
+			if err := key.PublicKey.Validate(); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func GetSigningJwk(config *JWTConfiguration) (*jwk.Key, error) {
+	if config.Keys == nil {
+		symmetricJwk, err := getSymmetricKey(config)
+		if err != nil {
+			return nil, err
+		}
+		return symmetricJwk, nil
+	}
+	for _, key := range config.Keys {
+		if key.PrivateKey.KeyUsage() == "enc" {
+			return &key.PrivateKey, nil
+		}
+	}
+	// return symmetric secret anyway to preserve existing behavior
+	symmetricJwk, err := getSymmetricKey(config)
+	if err != nil {
+		return nil, err
+	}
+	return symmetricJwk, nil
+}
+
+func GetSigningKey(k *jwk.Key) (any, error) {
+	var key any
+	switch (*k).KeyType() {
+	case jwa.OctetSeq:
+		key = []byte{}
+	case jwa.EC:
+		key = ecdh.PrivateKey{}
+	case jwa.RSA:
+		key = rsa.PrivateKey{}
+	case jwa.OKP:
+		// OKP is used for EdDSA keys
+		key = ed25519.PrivateKey{}
+	case jwa.InvalidKeyType:
+		return nil, jwt.ErrInvalidKeyType
+	}
+	if err := (*k).Raw(&key); err != nil {
+		return nil, err
+	}
+	return key, nil
+}
+
+func GetSigningAlg(k *jwk.Key) jwt.SigningMethod {
+	if k == nil {
+		return jwt.SigningMethodHS256
+	}
+
+	switch (*k).Algorithm().String() {
+	case "RS256":
+		return jwt.SigningMethodRS256
+	case "RS512":
+		return jwt.SigningMethodRS512
+	case "ES256":
+		return jwt.SigningMethodES256
+	case "ES512":
+		return jwt.SigningMethodES512
+	case "EdDSA":
+		return jwt.SigningMethodEdDSA
+	}
+
+	// return HS256 to preserve existing behaviour
+	return jwt.SigningMethodHS256
+}
+
+func GetKeyByKid(kid string, config *JWTConfiguration) (any, error) {
```

Comment:
Probably better to call it `FindPublicKeyByID` or something.

Reply:
One more thing about this function. Note that with this implementation you're limiting using only one symmetric key for verification, which is an artificial limitation. So probably you shouldn't return the public key but rather the key you find and leave that to be verified in the JWT library.

Reply:
yeah i think we just need to reorder the checks, the idea is to have a safety net for backward compatibility when `GOTRUE_JWT_KEYS` is not set and use `GOTRUE_JWT_SECRET`